src/meadowrun/pip_integration.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+import asyncio
+import pkgutil
+import sys
+
+
+async def _pip_freeze(python_interpreter: str) -> str:
+    """Effectively returns the output of <python_interpreter> -m pip freeze"""
+
+    # `pip freeze` and `pip list --format=freeze` usually give identical results, but
+    # not for "packages installed via Direct URLs":
+    # https://github.com/pypa/pip/issues/8176 which poetry uses to install all packages.
+    # So for us, `pip list --format=freeze` gives us the output we want both with poetry
+    # and in a "regular" virtualenv. We should keep an eye on that issue, though, as it
+    # looks like this difference in behavior was not totally intentional and may change
+    # in the future.
+    p = await asyncio.create_subprocess_exec(
+        python_interpreter,
+        *("-m", "pip", "list", "--format=freeze"),
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+    )
+    stdout, stderr = await p.communicate()
+
+    if p.returncode != 0:
+        raise ValueError(
+            f"pip freeze failed, return code {p.returncode}: {stderr.decode()}"
+        )
+
+    return stdout.decode()
+
+
+async def pip_freeze_portable_other_interpreter(python_interpreter: str) -> str:
+    """
+    Returns the result of calling pip freeze on the specified python_interpreter, but
+    tries to exclude packages that won't be available on PyPI or otherwise won't be
+    available on a different machine (i.e. not portable).
+    """
+    # TODO we have not actually implemented the portable part of this yet
```

Comment:
Raise exn instead if the platforms don't match given low chances this will work?

The "tries to exclude packages" comment - I don't think that will work (except for the source installs). Hard to believe we can just get away with excluding things.

Reply:
This isn't for cross-platform compatibility, this is just for excluding the editable installs